### PR TITLE
chore(transaction): Use PhasedBarrier for easier synchronization

### PR DESCRIPTION
**Pull request by dranikpg**
This PR introduces a PhasedBarrier primitive to replace the manual control of callback dispatches within transaction. A separate primitive makes way for simpler and easier code, as well as more clearly defined state semantics.

This PR has no functional changes.

### CODE REVIEW

File: src/server/transaction.cc [hunk 1]
```diff
@@ -71,12 +71,61 @@ void RecordTxScheduleStats(const Transaction* tx) {
...
+bool Transaction::PhasedBarrier::Active() const {
+  return count_.load(memory_order_acquire) > 0;
+}
+
+void Transaction::PhasedBarrier::Dec(Transaction* keep_alive) {
+  // Prevent transaction from being destroyed after atomic was decreased and Wait() unlocked,
+  // but before foreign thread finished notifying.
+  // ::boost::intrusive_ptr guard(keep_alive);
+  CHECK_GT(keep_alive->GetUseCount(), 1u);  // Why would we call it without use_count??
+
```
Comment (dranikpg) [inline src/server/transaction.cc:108]: @romange Do I understand correctly, that with a long tx queue, it might occur that our poll jobs finish _before_ we execute, so we're hanging on WaitForShardCallbacks() with use_count = 1 and once we pass over this barrier, we can drop ourselves?

Because it somewhat feels artifical to me that we separately have use count and run count, where run_count should clearly be a subset of use_count, i.e. use_count should be twice higher, representing any potential referrer that can get stuck (either in the shard queue or in the tx queue)

But I can also keep the guard because it's simpler 🙂

File: src/server/transaction.h [hunk 2]
```diff
@@ -499,25 +513,10 @@ class Transaction {
...
 
-  void WaitForShardCallbacks() {
-    run_ec_.await([this] { return 0 == run_count_.load(std::memory_order_relaxed); });
-
-    // no reads after this fence will be reordered before it, and if a store operation sequenced
-    // before some release operation that happened before the fence in another thread, this store
-    // will be visible after the fence.
-    // In this specific case we synchronize with DecreaseRunCnt that releases run_count_.
-    // See #997 before changing it.
-    std::atomic_thread_fence(std::memory_order_acquire);
```
Comment (dranikpg) [inline src/server/transaction.h:510]: Isn't 
```c++
atomic.load(relaxed)
fence(acquire)
```

the same as just?
```c++
atomic.load(acquire)
```

I assume that using fences directly makes only sense when optimizing batch operations around atomics (i.e. lot's of operations before a single message passing)

Or does it make a difference because we run `load()` withing a ec_ more than once? (almost always twice)🤓 I suppose it doesn't have any impact because it's mostly a compiler fence here and we don't have many reads around

File: src/server/transaction.h [hunk 3]
```diff
@@ -433,6 +429,24 @@ class Transaction {
...
+
+    bool IsMyPhase(uint32_t phase) const;  // Assumes already Acquired. Return true if same phase
+    uint32_t DEBUG_Get() const;
+
+   private:
+    std::atomic_uint32_t phase_{0};
+    std::atomic_uint32_t count_{0};
+    EventCount ec_{};
+  };
+
```
Comment (dranikpg) [inline src/server/transaction.h:449]: My take is that extracting something into a simple abstraction (but with a clearly defined interface) makes the code much easier to reason about, because we now think in terms of the 4 operations, instead of raw atomics and events, mixed with regular code

Comment (dranikpg) [reply src/server/transaction.h:449]: It's sad we can't reuse helio's BlockingCounter for this. We could, but we have to get rid of heap allocation and allow peeking with acquire semantics... Which is more complicated than counting manually

File: src/server/transaction.cc [hunk 4]
```diff
@@ -847,90 +875,60 @@ void Transaction::Execute(RunnableType cb, bool conclude) {
...
+    EngineShard* shard = EngineShard::tlocal();
+    if (IsArmedInShard(shard->shard_id())) {
+      if (run_barrier_.IsMyPhase(phase)) {
+        // PollExecution doesn't necessarily this transaction, so execution and cleanup code belongs
+        // to RunInShard
         shard->PollExecution("exec_cb", this);
       } else {
         VLOG(1) << "Skipping PollExecution " << DebugId() << " sid(" << shard->shard_id() << ")";
       }
     }
```
Comment (dranikpg) [inline src/server/transaction.cc:925]: My radical opinion: is_armed doesn't have to be a separate atomic and can be a regular shard flag.

How do we read safely from it without exiting the execution phase at the same time?

1. Atomically increment `use_count` and enter the section if it's larger than 1
2. Because we added an artifical job, Wait() is guaranteed to be waiting for us
3. Safely read is_armed and determine here whether we need to run -> profit

Same logic applies to the phase, it's changed only after the execution phase, so it's safe to read while inside it

File: src/server/transaction.cc [hunk 5]
```diff
@@ -847,90 +875,60 @@ void Transaction::Execute(RunnableType cb, bool conclude) {
...
 
   auto* ss = ServerState::tlocal();
   if (unique_shard_cnt_ == 1 && ss->thread_index() == unique_shard_id_ &&
       ss->AllowInlineScheduling()) {
     DVLOG(1) << "Short-circuit ExecuteAsync " << DebugId();
     EngineShard::tlocal()->PollExecution("exec_cb", this);
-    intrusive_ptr_release(this);  // against use_count_.fetch_add above.
+    run_barrier_.Wait();  // End phase
     return;
   }
```
Comment (dranikpg) [inline src/server/transaction.cc:904]: This was a "bug" before (more like inconsistency that should've been DCHECKED), we kept run_count > 0. Also we don't really need use_count here

File: src/server/transaction.cc [hunk 6]
```diff
@@ -847,90 +875,60 @@ void Transaction::Execute(RunnableType cb, bool conclude) {
...
-      // now this callback from the previous operation finally runs and calls PollExecution.
-      // It is usually ok, but for single shard operations we abuse index 0 in shard_data_
-      // Therefore we may end up with a situation where this old callback runs on shard 7,
-      // accessing shard_data_[0] that now represents shard 5 for the next operation.
-      // seqlock provides protection for that so each cb will only run on the operation it has
-      // been tasked with.
-      // We also must first check is_armed and only then seqlock. The first check ensures that
-      // the coordinator thread crossed
-      // "run_count_.store(unique_shard_cnt_, memory_order_release);" barrier and our seqlock_
-      // is valid.
```
Comment (dranikpg) [inline src/server/transaction.cc:917]: Wow, we actually don't "abuse" shard 0 😮

File: src/server/transaction.h [hunk 7]
```diff
@@ -552,17 +553,20 @@ class Transaction {
...
-  // Never access directly with index, always use SidToId.
-  absl::InlinedVector<PerShardData, 4> shard_data_;  // length = shard_count
+  // Main synchronization point for dispatching hop callbacks and waiting for them to finish.
+  // After scheduling, sequential hops are executed as follows:
+  //    coordinator: Prepare hop, then Start(num_shards), dispatch poll jobs and Wait()
+  //    tx queue:    Once IsArmedInShard() /* checks Active() */ -> run in shard and Dec()
+  // As long as barrier is active any writes by the coordinator are prohibited, so shard threads
+  // can safely read transaction state and modify per-shard state belonging to them.
+  // Inter-thread synchronization is provided by the barriers acquire/release pairs.
+  PhasedBarrier run_barrier_;
```
Comment (dranikpg) [inline src/server/transaction.h:563]: Now our inter-thread coordination during hops is **clearly** defined in a single place in simple terms

File: src/server/transaction.cc [hunk 8]
```diff
@@ -847,90 +853,60 @@ void Transaction::Execute(RunnableType cb, bool conclude) {
...
-      if (seq_after == seq) {
-        // shard->PollExecution(this) does not necessarily execute this transaction.
-        // Therefore, everything that should be handled during the callback execution
-        // should go into RunInShard.
+    EngineShard* shard = EngineShard::tlocal();
+    if (IsArmedInShard(shard->shard_id())) {
+      if (run_barrier_.IsMyPhase(phase)) {
```
Comment (dranikpg) [inline src/server/transaction.cc:896]: Actually, why do we need the seqlock at all if we check `IsArmedInShard()` also for direct ooo runs in PollExecution? We'll silently skip it and it won't run twice 🤔

File: src/server/transaction.h [hunk 9]
```diff
@@ -433,6 +431,24 @@ class Transaction {
     }
   };
 
+  // Barrier akin to helio's BlockingCounter, but with phase counting and proper acquire semantics
+  // for polling work from other threads. And without keap allocation.
```
Comment (romange) [inline src/server/transaction.h:435]: keap -> heap

File: src/server/transaction.h [hunk 10]
```diff
@@ -433,6 +431,24 @@ class Transaction {
...
+  struct PhasedBarrier {
```
Comment (romange) [inline src/server/transaction.h:436]: class

File: src/server/transaction.h [hunk 11]
```diff
@@ -433,6 +431,24 @@ class Transaction {
...
+    uint32_t Start(uint32_t count);  // Release: Store count, return current phase
```
Comment (romange) [inline src/server/transaction.h:437]: phase is like seqlock? 
add a comment that it corresponds to a specific hop in a transaction

File: src/server/transaction.h [hunk 12]
```diff
@@ -433,6 +431,24 @@ class Transaction {
...
+    bool Active() const;             // Acquire: Return if count > 0. Use for polling for work
+
+    void Dec(Transaction* keep_alive);  // Relase: Decrease count, notify ec on 0
+    void Wait();                        // Acquire: Wait until count = 0, increment phase
+
+    bool IsMyPhase(uint32_t phase) const;  //  Return if same phase. Assumes called after Active().
+    uint32_t DEBUG_Get() const;
```
Comment (romange) [inline src/server/transaction.h:444]: what DEBUG_Get returns?

File: src/server/transaction.h (hunk 2 above)
Comment (romange) [reply src/server/transaction.h:510]: atomic_thread_fence is not a compiler fence

File: src/server/transaction.h [hunk 13]
```diff
@@ -585,21 +589,16 @@ class Transaction {
   DbIndex db_index_{0};
   uint64_t time_now_ms_{0};
 
-  std::atomic_uint32_t wakeup_requested_{0};  // whether tx was woken up
-  std::atomic_uint32_t use_count_{0}, run_count_{0}, seqlock_{0};
+  std::atomic_uint32_t use_count_{0};  // transaction exists only as a shared pointer (intrusive)
```
Comment (romange) [inline src/server/transaction.h:592]:
```suggestion
  std::atomic_uint32_t use_count_{0};  // transaction exists only as an intrusive_ptr
```

File: src/server/transaction.cc [hunk 14]
```diff
@@ -243,26 +289,6 @@ void Transaction::StoreKeysInArgs(KeyIndex key_index, bool rev_mapping) {
...
- *    In that case multiple threads will be scheduled. Similarly they have a schedule barrier,
- *    and IsArmedInShard can read any variable from shard_data[x].
- * c. Trans spans a single shard and it's multi. shard_data has size of ess_.size.
- *    IsArmedInShard will check shard_data[x].
- * d. Trans spans multiple shards and it's multi. Similarly shard_data[x] will be checked.
- *    unique_shard_cnt_ and unique_shard_id_ are not accessed until shard_data[x] is armed, hence
- *    we have a barrier between coordinator and engine-threads. Therefore there should not be
- *    data races.
- *
- **/
```
Comment (romange) [inline src/server/transaction.cc:264]: why did you delete this comment?

File: src/server/transaction.cc [hunk 15]
```diff
@@ -794,9 +803,6 @@ void Transaction::UnlockMulti() {
   unsigned shard_journals_cnt =
       ServerState::tlocal()->journal() ? CalcMultiNumOfShardJournals() : 0;
 
-  uint32_t prev = run_count_.fetch_add(shard_data_.size(), memory_order_relaxed);
```
Comment (romange) [inline src/server/transaction.cc:797]: what happens here?

File: src/server/transaction.cc [hunk 16]
```diff
@@ -71,12 +71,58 @@ void RecordTxScheduleStats(const Transaction* tx) {
...
+  ::boost::intrusive_ptr guard(keep_alive);
+
+  uint32_t before = count_.fetch_sub(1);
+  CHECK_GE(before, 1u);
+  if (before == 1)
+    ec_.notify();
+}
+
+void Transaction::PhasedBarrier::Wait() {
+  ec_.await([this] { return count_.load(memory_order_acquire) == 0; });
```
Comment (romange) [inline src/server/transaction.cc:113]: it's a bug and very subtle one. You can not do this:
```
Tthread1:  cond = true; ec->notify();
Thread2: ec->await(cond); .... delete ec;
```

There are several reasons, one of them thread1 can execute `cond=true`, then thread2 will execute await, continue deleting ec and then thread1 will have data race.

Even more subtle, notify itself wakes up thread2 but it still references ec one line after, so it can be and actually happens that thread 2 wakes up, deletes ec and then thread1 has a data race,

This is why Done wraps eventcount with reference count

**Review by romange — CHANGES_REQUESTED** [review]

File: src/server/transaction.h (hunk 2 above)
Comment (dranikpg) [reply src/server/transaction.h:510]: Every atomic fence is a compiler fence, on x86 if often actually nothing more than than 🙂

> On x86 (including x86-64), atomic_thread_fence functions issue no CPU instructions and only affect compile-time code motion, except for std::atomic_thread_fence(std::memory_order::seq_cst)

https://en.cppreference.com/w/cpp/atomic/atomic_thread_fence

File: src/server/transaction.h (hunk 11 above)
Comment (dranikpg) [reply src/server/transaction.h:437]: phase is seqlock, I added a comment in transaction.h that a phase is a hop

File: src/server/transaction.cc (hunk 15 above)
Comment (dranikpg) [reply src/server/transaction.cc:797]: We don't need it https://github.com/dragonflydb/dragonfly/pull/2455#discussion_r1461804316. We never use this value again, except for CHECKs

File: src/server/transaction.cc (hunk 16 above)
Comment (dranikpg) [reply src/server/transaction.cc:113]: This is why I asked https://github.com/dragonflydb/dragonfly/pull/2455#discussion_r1461749346

It's the same code, but wrapped in a struct. I don't agree there is a bug here, because we still rely on use_count. 

I agree it's tricky. Once our guard from `Dec()` drops, the object can be gone, _while_ we're exiting the function. But we're been doing the same thing before with `DecreaseRunCnt()`

Comment (dranikpg) [reply src/server/transaction.cc:113]: I can refactor it into a static function so it's explicit that the object it not accessed anymore, but we can still be deleted from inside the transaction when we call `Dec()` instead of `DecreaseRunCnt()`

File: src/server/transaction.cc (hunk 1 above)
Comment (romange) [reply src/server/transaction.cc]: yes, you are correct. by shard queue  you mean `fb2::FiberQueue` ?

File: src/server/transaction.cc [hunk 17]
```diff
@@ -71,12 +71,58 @@ void RecordTxScheduleStats(const Transaction* tx) {
...
+}
+
+bool Transaction::PhasedBarrier::Active() const {
+  return count_.load(memory_order_acquire) > 0;
+}
+
+void Transaction::PhasedBarrier::Dec(Transaction* keep_alive) {
+  // Prevent transaction from being destroyed after count was decreased and Wait() unlocked,
+  // but before this thread finished notifying.
+  ::boost::intrusive_ptr guard(keep_alive);
```
Comment (romange) [inline src/server/transaction.cc:104]: nice. I do not know why i have not done it myself. I did it in other places.

File: src/server/transaction.cc (hunk 16 above)
Comment (dranikpg) [reply src/server/transaction.cc:113]: https://isocpp.org/wiki/faq/freestore-mgmt#delete-this

File: src/server/transaction.cc (hunk 14 above)
Comment (romange) [reply src/server/transaction.cc:264]: rewrite it then 😄

File: src/server/transaction.cc (hunk 16 above)
Comment (romange) [reply src/server/transaction.cc:113]: no need to refactor if it works well like this

File: src/server/transaction.h (hunk 11 above)
Comment (romange) [reply src/server/transaction.h:438]: not that I suggest doing it in this PR but did you consider merging phase and count into the same atomic_uint32?

File: src/server/transaction.h (hunk 2 above)
Comment (romange) [reply src/server/transaction.h:510]: what you said, I wanted to avoid acquire load inside event count callback. probably over-optimization

File: src/server/transaction.cc [hunk 18]
```diff
@@ -71,12 +71,58 @@ void RecordTxScheduleStats(const Transaction* tx) {
...
+  ec_.await([this] { return count_.load(memory_order_acquire) == 0; });
+  phase_.fetch_add(1, memory_order_relaxed);
+}
+
+bool Transaction::PhasedBarrier::IsMyPhase(uint32_t phase) const {
+  DCHECK_GT(DEBUG_Get(), 0u);
+  return phase_.load(memory_order_relaxed) == phase;
+}
+
+uint32_t Transaction::PhasedBarrier::DEBUG_Get() const {
```
Comment (romange) [inline src/server/transaction.cc:122]: nit: DEBUG_Count is clearer imho

File: src/server/transaction.cc (hunk 8 above)
Comment (romange) [reply src/server/transaction.cc:896]: you deleted the comment below that explains why we need to check the seqlock. I would like to restore it unless you think it's wrong

**Review by romange — COMMENTED** [review]
Interesting PR. Please add to PR description not only the motivation but also the description of major changes your performed. It took me a while to understand that state variables have not changed, they just got renamed (or maybe I am wrong and then you can clear this up)

File: src/server/transaction.h (hunk 11 above)
Comment (dranikpg) [reply src/server/transaction.h:438]: No need, because #2457 will get rid of it at all

File: src/server/transaction.cc [hunk 19]
```diff
@@ -709,71 +726,56 @@ OpStatus Transaction::ScheduleSingleHop(RunnableType cb) {
...
-    // IsArmedInShard() first checks run_count_ before shard_data, so use release ordering.
-    shard_data_[SidToId(unique_shard_id_)].is_armed.store(true, memory_order_relaxed);
-    run_count_.store(1, memory_order_release);
-
     time_now_ms_ = GetCurrentTimeMs();
+    shard_data_[SidToId(unique_shard_id_)].is_armed.store(true, memory_order_relaxed);
 
-    // NOTE: schedule_cb cannot update data on stack when run_fast is false.
-    // This is because ScheduleSingleHop can finish before the callback returns.
+    // Start new phase, be careful with writes until phase end!
```
Comment (romange) [inline src/server/transaction.cc:743]: "be careful with writes" is not clear :)
what do you mean?

**Comment by romange** [comment]
Looks good, added minor comments. Also, please add to the PR description whether this PR has functional changes and if it has - where and I will double check. I suggest running this PR on sidekiq benchmark + regression tests too

**Comment by dranikpg** [comment]
Ran the loadtest 10 times, seems to work

File: src/server/transaction.cc (hunk 19 above)
Comment (dranikpg) [reply src/server/transaction.cc:752]: Just to be careful 😄 There _are_ all kinds of writes below, so we can't say they're not allowed, it's just based on the single shard logic
